New module: Add module or managing Windows Active Directory users (windows/win_domain_user)

nwchandler opened the pull request:
##### SUMMARY
This PR adds support for managing Windows Active Directory users with Ansible. The existing win_user module is designed to support management of local users. There are enough differences between local user management and domain user management that a separate module seemed to make sense. See discussion on Issue #20428. 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
windows/win_domain_user.ps1

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Jun 29 2016, 14:05:02) [GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
```


##### ADDITIONAL INFORMATION

nwchandler committed:
Initial win_domain_user module support

nwchandler committed:
Add return information

nwchandler committed:
Update return values

jhawkesworth commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,272 @@
+# POWERSHELL_COMMON
+
+########
+Import-Module ActiveDirectory
```
I think this line will fail if run against a non-active directory server.  Consider a try catch and fail-json if the ActiveDirectory module is not available.

nwchandler replied:
Thanks, for the review! Good idea. I will make the update.

nwchandler committed:
Add try/catch for PS module import

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+########
+
+$result = @{
+    changed = $false;
```
Trailing semi-colon is not needed.

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+    Fail-Json $result "Failed to import ActiveDirectory PowerShell module. This module should be run on a domain controller, and the ActiveDirectory module must be available."
+}
+
+$params = Parse-Args $args
```
Please look into adding check-mode support. Often you can get away with some strategically places `-WhatIf:$check_mode` statements !

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+      - When C(present), creates or updates the user account.  When C(absent),
+        removes the user account if it exists.  When C(query),
+        retrieves the user account details without making any changes.
+    required: false
```
If parameters are not required, you don't have to add `required: false`. Only add `required: true` if they are required.

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+      - C(no) will unlock the user account if locked.
+    required: false
+    choices: [ 'no' ]
+    default: null
```
This is not needed, the default value is null, so it's implicit.

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+    description:
+      - C(no) will unlock the user account if locked.
+    required: false
+    choices: [ 'no' ]
```
Wouldn't the choice be [ 'yes', 'no' ] as you'd expect from a boolean ?
As per previous remark, this would become:
```yaml
    type: bool
```

jborean93 commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+}
+
+# Parameter validation
+If ($account_locked -ne $null -and $account_locked) {
```
My brain isn't working right now are you able to explain this logic? To me it seems like it will fail if `$account_locked` is set to `true` if it has been set? Why can't we create a locked account?

nwchandler replied:
I followed Microsoft's lead on this one. There is an Unlock-ADAccount cmdlet, but no corresponding Lock-ADAccount. The thinking is that an account would be locked because a user tried to login too many times with incorrect credentials, so an admin would unlock an account, but not lock it. If an account needed to be "locked" by an admin, he/she would just disable the account by setting "enabled" to false.

jborean93 replied:
No worries that makes sense, I can see the same functionality in the win_user module this was just me being ignorant :)

jborean93 commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+        If ($password -and (($new_user -and $update_password -eq "on_create") -or $update_password -eq "always")) {
+            $secure_password = ConvertTo-SecureString $password -AsPlainText -Force
+            Set-ADAccountPassword -Identity $username -Reset:$true -Confirm:$false -NewPassword $secure_password
+            $user_obj = Get-ADUser -Identity $username -Properties *
```
Should we just run this once after creating the user to save getting these details everytime we want to make a change?

nwchandler replied:
That's what I did initially, but I found problems in testing. As accounts were updated, I ran into problems where later logic would fail. I ended up deciding it was better to refresh the state of the user object after every update.
 (Sorry, I know that's vague; I don't recall the specific issues I ran into off-hand.)

jborean93 replied:
Fair enough, sounds like a typical Microsoft thing to me.

jborean93 commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+$password_never_expires = Get-AnsibleParam -obj $params -name "password_never_expires" -type "bool"
+$user_cannot_change_password = Get-AnsibleParam -obj $params -name "user_cannot_change_password" -type "bool"
+$account_locked = Get-AnsibleParam -obj $params -name "account_locked" -type "bool"
+$groups = Get-AnsibleParam -obj $params -name "groups"
```
There is now a `list` type that will automatically convert a the item to a list if it isn't already.

nwchandler replied:
Hi, @jborean93 - My apologies for the late reply, but thank you for the comments. Good to know about the list type; I'll update this.

jborean93 commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,278 @@
+
+        # Configure group assignment
+        If ($groups -ne $null) {
+            If ($groups -is [System.String]) {
```
If using the `list` type it will automatically convert it to a list and split it by `,` if it is a string.

jborean93 commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+
+
+DOCUMENTATION = r'''
+---
```
Can you add a notes section around where this module can run and the requirements?

nwchandler replied:
Yes, I'll add that.

jborean93 reviewed (changes requested):
Looks like a great module to have, just a few questions and queries.

nathanwebsterdotme commented:
Looking forward to this one!

dagwieers commented:
@nathanwebsterdotme If you do, I'd suggest you test it out in your environment and provide feedback. And help with code-review. Those things will speed up it being merged.

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+    description:
+      - C(yes) will enable the user account. C(no) will disable the account.
+    required: false
+    choices:
```
Nowadays we do:
```yaml
    type: bool
    default: 'yes'
```

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+version_added: "2.4"
+short_description: Manages Windows Active Directory user accounts
+description:
+     - Manages Windows Active Directory user accounts
```
Trailing dot required. I know, quite pedantic rules and all.

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,361 @@
+    default: null
+
+author:
+    - "Nick Chandler (@nwchandler)"
```
We tend to not add quotes where they are no needed in YAML.
Here they are not needed.

nathanwebsterdotme commented:
I've been testing this in conjunction with the win_domain module and am getting some issues, thought I'd note them down here. 

My playbook basically builds a domain controller from scratch on AWS using the following steps: 

1. Deploy AWS instance
2. Set windows-hostname (and restart the machine)
3. Create domain & dcpromo the instance (using win_domain) + restart
4. Create a Domain Admin user. 

No matter what I've tried, I get the following error from this module: 
`"msg": "Unable to find a default server with Active Directory Web Services running.",`

I event added in a task to ensure the ADWS service is running prior to trying to add the domain user, and still, it reports this message.

Interestingly, if I run this module against the very same box as an individual task immediately after the failure, it works OK.

jborean93 commented:
Hey @nwchandler I'll be keen to try and get this in before the next 2.4 release, do you think you will have time to try and incorporate the changes requested?

If not I'm happy to work on it and try and get it in sometime soon.

dagwieers commented:
+label new_module

nwchandler commented:
Sorry for the slow responses, all. I think I should have an opportunity to make the requested changes by early next week if that is okay.

Commit:
Improve win_domain_user module

Commit:
Fix bad merge

Commit:
Fix pep8 failure

nwchandler commented:
Hi, all. I believe I've addressed the comments in the new version of code (pending one last Shippable test). @dagwieers - I did some work on enabling check_mode, but it gave me some problems. I think it may take a little more re-work than I originally envisioned. So, I pushed the other requested changes. I can take another swing at it at some point, but it may be a while before I can, and I wanted to get the other updates in. I'll defer to you whether that would be a blocker or would be something we could add in the future.

Commit:
Actually fix pep8 failure

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,332 @@
+DOCUMENTATION = r'''
+---
+module: win_domain_user
+version_added: 2.4
```
This should be a string:
```yaml
version_added: '2.4'
```

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,332 @@
+      - C(yes) will require the user to change their password at next login.
+        C(no) will clear the expired password flag. This is mutually exclusive
+        with I(password_never_expires).
+    choices: [ 'yes', 'no' ]
```
This should be:
```yaml
type: bool
```

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,332 @@
+    description:
+      - C(yes) will set the password to never expire.  C(no) will allow the
+        password to expire. This is mutually exclusive with I(password_expired)
+    choices: [ 'yes', 'no' ]
```
This should be:
```yaml
type: bool
```

dagwieers commented on lib/ansible/modules/windows/win_domain_user.py:
```diff
@@ -0,0 +1,332 @@
+    description:
+      - C(yes) will prevent the user from changing their password.  C(no) will
+        allow the user to change their password.
+    choices: [ 'yes', 'no' ]
```
This should be:
```yaml
type: bool
```

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+        # If the account does not exist, create it
+        If (-not $user_obj) {
+            If ($path -ne $null){
+                New-ADUser -Name $username -Path $path
```
Add here `-WhatIf:$check_mode`

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+                New-ADUser -Name $username -Path $path
+            }
+            Else {
+                New-ADUser -Name $username
```
Add here `-WhatIf:$check_mode`

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+            Else {
+                New-ADUser -Name $username
+            }
+            $user_obj = Get-ADUser -Identity $username -Properties *
```
I guess, skip everything that follows on $check_mode (e.g. by running Exit-Json)

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+
+        # Configure password policies
+        If (($password_never_expires -ne $null) -and ($password_never_expires -ne $user_obj.PasswordNeverExpires)) {
+            Set-ADUser -Identity $username -PasswordNeverExpires $password_never_expires
```
Add here `-WhatIf:$check_mode`

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+            $result.changed = $true
+        }
+        If (($password_expired -ne $null) -and ($password_expired -ne $user_obj.PasswordExpired)) {
+            Set-ADUser -Identity $username -ChangePasswordAtLogon $password_expired
```
Add here `-WhatIf:$check_mode`

dagwieers commented on lib/ansible/modules/windows/win_domain_user.ps1:
```diff
@@ -0,0 +1,271 @@
+            $result.changed = $true
+        }
+        If (($user_cannot_change_password -ne $null) -and ($user_cannot_change_password -ne $user_obj.CannotChangePassword)) {
+            Set-ADUser -Identity $username -CannotChangePassword $user_cannot_change_password
```
Add here `-WhatIf:$check_mode`

dagwieers commented:
@nwchandler check-mode is a must-have for every (new) module. It is best to first make it full-featured, then look at supporting check-mode. In this case, I think you can get away with adding `-WhatIf:$check_mode` to most statements that modify things (like Add-ADUser, Set-ADUser, ...) and leave out some of the post-change validation/lookups. Maybe terminate the module prematurely if needed (even if that means not everything can be checked properly). Check-mode is often best-effort (and heavily depends on low-level implementation details we have no control over).

Let me know if you need help.

Commit:
Update win_domain_user.py to meet standards

Commit:
Add check_mode support for win_domain_user

nwchandler commented:
Thanks for the feedback and ideas on check_mode support, @dagwieers. I have made some updates accordingly. Short-circuiting and exiting early is a good idea.

jborean93 reviewed (approved):
Looks good, just did some local tests and things seem to work correctly here. Keen to get this in and people testing it in real life situations.

jborean93 commented:
shitpit

nwchandler commented:
Thanks for everyone's reviews and feedback!

dagwieers commented:
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

nwchandler commented:
It looks like @jborean93 approved - anything else I need to do to make this happen, @dagwieers ?

dagwieers commented:
@nwchandler If you're asking me specifically, the only thing I noted was the use of `type: bool` in the documentation. Ah, and now I see you did add it, but it should replace the `choices: ['no', 'yes']` part.

jborean93 committed:
Updated documentation before merge

jborean93 commented:
Thanks for the work here @nwchandler, I'm going to merge this in so people can start using it in devel. Are you able to update https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml and https://github.com/ansible/ansible/blob/devel/CHANGELOG.md with the relevant info for the new module.

dagwieers commented:
BOTMETA should not require updating if the author(s) are listed in the module.

jborean93 commented:
oh did not know that, the more you know
